api/post.go

```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
 	var rpost *model.Post
 	if result := <-Srv.Store.Post().Save(post); result.Err != nil {
 		return nil, result.Err
-	} else if doUpdateLastViewed && (<-Srv.Store.Channel().UpdateLastViewedAt(post.ChannelId, c.Session.UserId)).Err != nil {
-		return nil, result.Err
 	} else {
 		rpost = result.Data.(*model.Post)
 
 		fireAndForgetNotifications(rpost, c.Session.TeamId, c.GetSiteURL())
 
+		if triggerWebhooks {
+			fireAndForgetWebhookEvent(c, rpost)
+		}
+
 	}
 
 	return rpost, nil
 }
 
+func fireAndForgetWebhookEvent(c *Context, post *model.Post) {
+
+	go func() {
+
+		chchan := Srv.Store.Webhook().GetOutgoingByChannel(post.ChannelId)
+
+		firstWord := strings.Split(post.Message, " ")[0]
+		thchan := Srv.Store.Webhook().GetOutgoingByTriggerWord(c.Session.TeamId, post.ChannelId, firstWord)
+
+		hooks := []*model.OutgoingWebhook{}
+
+		if result := <-chchan; result.Err != nil {
+			l4g.Error("Encountered error getting webhook by channel, err=%v", result.Err)
+			return
+		} else {
+			hooks = append(hooks, result.Data.([]*model.OutgoingWebhook)...)
+		}
+
+		if result := <-thchan; result.Err != nil {
+			l4g.Error("Encountered error getting webhook by trigger word, err=%v", result.Err)
+			return
+		} else {
+			hooks = append(hooks, result.Data.([]*model.OutgoingWebhook)...)
+		}
+
+		cchan := Srv.Store.Channel().Get(post.ChannelId)
+		uchan := Srv.Store.User().Get(post.UserId)
+
+		tchan := make(chan *model.Team)
+		isTeamBeingFetched := make(map[string]bool)
+		for _, hook := range hooks {
+			if !isTeamBeingFetched[hook.TeamId] {
+				isTeamBeingFetched[hook.TeamId] = true
+				go func() {
+					if result := <-Srv.Store.Team().Get(hook.TeamId); result.Err != nil {
+						l4g.Error("Encountered error getting team, team_id=%s, err=%v", hook.TeamId, result.Err)
+						tchan <- nil
+					} else {
+						tchan <- result.Data.(*model.Team)
+					}
+				}()
+			}
+		}
+
+		teams := make(map[string]*model.Team)
+		for range isTeamBeingFetched {
+			team := <-tchan
+			if team != nil {
+				teams[team.Id] = team
+			}
+		}
+
+		var channel *model.Channel
+		if result := <-cchan; result.Err != nil {
+			l4g.Error("Encountered error getting channel, channel_id=%s, err=%v", post.ChannelId, result.Err)
```

Comment:
This is on purpose since we can still post the message without the channel (we just pull it for the name). This is rarely going to fail since the channel must exist as it's confirmed in createPost. 0/5 if you a return here anyways